proxy-socket-core/src/main/java/net/airvantage/proxysocket/core/v2/ProxyProtocolV2Decoder.java

```diff
@@ -0,0 +1,231 @@
+        };
+    }
+
+    private static int IPV4_ADDR_PAIR_LEN = 2*(IPV4_ADDR_LEN + PORT_LEN);
+    private static AddressPair parseIPv4Addresses(byte[] data, int pos, int variableLength)
+            throws ProxyProtocolParseException {
+        if (variableLength < IPV4_ADDR_PAIR_LEN) {
+            throw new ProxyProtocolParseException("Truncated IPv4 address block in header");
+        }
+
+        InetAddress s;
+        InetAddress d;
+        try {
+            s = InetAddress.getByAddress(new byte[]{data[pos++], data[pos++], data[pos++], data[pos++]});
+            d = InetAddress.getByAddress(new byte[]{data[pos++], data[pos++], data[pos++], data[pos++]});
+        } catch (UnknownHostException e) {
+            throw new ProxyProtocolParseException("Invalid IPv4 address in header", e);
+        }
+
+        int sp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+        int dp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+
+        return new AddressPair(new InetSocketAddress(s, sp), new InetSocketAddress(d, dp), IPV4_ADDR_PAIR_LEN);
+    }
+
+    private static int IPV6_ADDR_PAIR_LEN = 2*(IPV6_ADDR_LEN + PORT_LEN);
+    private static AddressPair parseIPv6Addresses(byte[] data, int pos, int variableLength)
+            throws ProxyProtocolParseException {
+        if (variableLength < IPV6_ADDR_PAIR_LEN) {
+            throw new ProxyProtocolParseException("Truncated IPv6 address block in header");
+        }
+
+        InetAddress s;
+        InetAddress d;
+        byte[] sb = new byte[IPV6_ADDR_LEN];
+        byte[] db = new byte[IPV6_ADDR_LEN];
+        System.arraycopy(data, pos, sb, 0, IPV6_ADDR_LEN);
+        System.arraycopy(data, pos+IPV6_ADDR_LEN, db, 0, IPV6_ADDR_LEN);
+        try {
+            s = InetAddress.getByAddress(sb);
+            d = InetAddress.getByAddress(db);
+        } catch (UnknownHostException e) {
+            throw new ProxyProtocolParseException("Invalid IPv6 address in header", e);
+        }
+
+        pos += 2*IPV6_ADDR_LEN;
+        int sp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+        int dp = ((data[pos++] & 0xFF) << 8) | (data[pos++] & 0xFF);
+
+        return new AddressPair(new InetSocketAddress(s, sp), new InetSocketAddress(d, dp), IPV6_ADDR_PAIR_LEN);
+    }
+
+    private static int UNIX_ADDR_PAIR_LEN = 2*UNIX_ADDR_LEN;
+    private static AddressPair parseUnixAddresses(byte[] data, int pos, int variableLength)
+            throws ProxyProtocolParseException {
+        if (variableLength < UNIX_ADDR_PAIR_LEN) {
+            throw new ProxyProtocolParseException("Truncated UNIX address block in header");
+        }
+
+        // A receiver is not required to implement other ones, provided that it
+        // automatically falls back to the UNSPEC mode for the valid combinations above
+        // that it does not support.
+        return new AddressPair(null, null, UNIX_ADDR_PAIR_LEN);
+    }
```

Comment:
I'm not sure what we want to do here?
But Reading the code I don't see what is returned will be consider as an UNSPEC or I missed something ? 🤔

Reply:
Ok I think I was maybe a bit confused by the comment

```java
// A receiver is not required to implement other ones, provided that it
// automatically falls back to the UNSPEC mode for the valid combinations above
// that it does not support.
```
Reading that I was espected that an UNSPEC header will be create.


I don't think this is what we should do but what the comment makes me believe.
I don't know what could be the right behaviour.
But current behavior is OK for a first PR.